go/x509util/x509util.go

```diff
@@ -0,0 +1,576 @@
+	}
+	addList := func(items []string, prefix string) {
+		for _, item := range items {
+			addSingle(item, prefix)
+		}
+	}
+	addList(name.Country, "C=")
+	addList(name.Organization, "O=")
+	addList(name.OrganizationalUnit, "OU=")
+	addList(name.Locality, "L=")
+	addList(name.Province, "ST=")
+	addList(name.StreetAddress, "streetAddress=")
+	addList(name.PostalCode, "postalCode=")
+	addSingle(name.SerialNumber, "serialNumber=")
+	addSingle(name.CommonName, "CN=")
+	for _, atv := range name.Names {
+		value, ok := atv.Value.(string)
+		if !ok {
+			continue
+		}
+		t := atv.Type
+		if len(t) == 4 && t[0] == OidAttribute[0] && t[1] == OidAttribute[1] && t[2] == OidAttribute[2] {
+			// OID is 'id-at N'
+			switch t[3] {
+			case OidCommonName[3], OidSerialNumber[3], OidCountry[3], OidLocality[3], OidProvince[3],
+				OidStreetAddress[3], OidOrganization[3], OidOrganizationalUnit[3], OidPostalCode[3]:
+				continue // covered by explicit fields
+			case OidPseudonym[3]:
+				addSingle(value, "pseudonym=")
+				continue
+			case OidTitle[3]:
+				addSingle(value, "title=")
+				continue
+			case OidDnQualifier[3]:
+				addSingle(value, "dnQualifier=")
+				continue
+			case OidName[3]:
+				addSingle(value, "name=")
+				continue
+			case OidSurname[3]:
+				addSingle(value, "surname=")
+				continue
+			case OidGivenName[3]:
+				addSingle(value, "givenName=")
+				continue
+			case OidInitials[3]:
+				addSingle(value, "initials=")
+				continue
+			case OidGenerationQualifier[3]:
+				addSingle(value, "generationQualifier=")
+				continue
+			}
+		}
+		addSingle(value, t.String()+"=")
+	}
+	return result.String()
+}
+
+// CertificateToString generates a string describing the given certificate.
+// The output roughly resembles that from openssl x509 -text.
+func CertificateToString(cert *x509.Certificate) string {
+	var result bytes.Buffer
+	result.WriteString(fmt.Sprintf("Certificate:\n"))
+	result.WriteString(fmt.Sprintf("    Data:\n"))
+	result.WriteString(fmt.Sprintf("        Version: %d (%#[1]x)\n", cert.Version))
+	result.WriteString(fmt.Sprintf("        Serial Number: %d (%#[1]x)\n", cert.SerialNumber))
+	result.WriteString(fmt.Sprintf("    Signature Algorithm: %v\n", cert.SignatureAlgorithm))
+	result.WriteString(fmt.Sprintf("        Issuer: %v\n", NameToString(cert.Issuer)))
+	result.WriteString(fmt.Sprintf("        Validity:\n"))
+	result.WriteString(fmt.Sprintf("            Not Before: %v\n", cert.NotBefore))
+	result.WriteString(fmt.Sprintf("            Not After : %v\n", cert.NotAfter))
+	result.WriteString(fmt.Sprintf("        Subject: %v\n", NameToString(cert.Subject)))
+	result.WriteString(fmt.Sprintf("        Subject Public Key Info:\n"))
+	result.WriteString(fmt.Sprintf("            Public Key Algorithm: %v\n", publicKeyAlgorithmToString(cert.PublicKeyAlgorithm)))
+	result.WriteString(fmt.Sprintf("%v\n", publicKeyToString(cert.PublicKeyAlgorithm, cert.PublicKey)))
+
+	var showCritical = func(critical bool) {
```

Comment:
Could you move this var definition down to just after the if statement (rather than just before), so that is closer to the place where it is first used?  Again, just for ease of reading :)